enzyme/Enzyme/DifferentialUseAnalysis.h

```diff
@@ -433,11 +465,68 @@ void minCut(const llvm::DataLayout &DL, llvm::LoopInfo &OrigLI,
             const llvm::SetVector<llvm::Value *> &Recomputes,
             const llvm::SetVector<llvm::Value *> &Intermediates,
             llvm::SetVector<llvm::Value *> &Required,
-            llvm::SetVector<llvm::Value *> &MinReq,
-            const llvm::ValueMap<llvm::Value *, GradientUtils::Rematerializer>
-                &rematerializableAllocations,
+            llvm::SetVector<llvm::Value *> &MinReq, const GradientUtils *gutils,
             llvm::TargetLibraryInfo &TLI);
 
+template <typename ft>
+__attribute__((always_inline)) static inline void
+forAllDirectUsers(ft &&f, const GradientUtils *gutils, llvm::Instruction *IVI,
+                  llvm::Value *val, bool useCheck) {
+  using namespace llvm;
+  if (!gutils->isConstantValue(IVI))
+    return;
+  bool inserted = false;
+  if (auto II = dyn_cast<InsertValueInst>(IVI))
+    inserted = II->getInsertedValueOperand() == val ||
+               II->getAggregateOperand() == val;
+  if (auto II = dyn_cast<ExtractValueInst>(IVI))
+    inserted = II->getAggregateOperand() == val;
+  if (auto II = dyn_cast<InsertElementInst>(IVI))
+    inserted = II->getOperand(1) == val || II->getOperand(0) == val;
+  if (auto II = dyn_cast<ExtractElementInst>(IVI))
+    inserted = II->getOperand(0) == val;
+  if (inserted) {
+    SmallVector<Instruction *, 1> todo;
+    todo.push_back(IVI);
+    while (todo.size()) {
+      auto cur = todo.pop_back_val();
+      for (auto u : cur->users()) {
+        if (isa<InsertValueInst>(u) || isa<InsertElementInst>(u) ||
+            isa<ExtractValueInst>(u) || isa<ExtractElementInst>(u)) {
+          auto I2 = cast<Instruction>(u);
+          bool subCheck = useCheck;
+          if (!subCheck) {
+            subCheck = is_value_needed_in_reverse<QueryType::Shadow>(
+                gutils, I2, gutils->mode, gutils->notForAnalysis);
+          }
+          if (subCheck)
+            f(I2);
+          todo.push_back(I2);
+          continue;
+        }
+      }
+    }
+  }
+}
+
+template <typename ft>
+__attribute__((always_inline)) static inline void
+forAllDifferentialUsersOf(ft &&f, const GradientUtils *gutils, llvm::Value *V,
```

Comment:
```suggestion
__attribute__((always_inline)) static inline void
forAllDifferentialUsersOf(llvm::function_ref<void(llvm::User*)>, const GradientUtils *gutils, llvm::Value *V,
```

templating does not seem to be needed here.